casper/src/main/scala/coop/rchain/casper/ReportingCasper.scala

```diff
@@ -295,43 +291,61 @@ object ReportingCasper {
+                                       processedDeploy.deploy
+                                     ).attempt
+        failureEither <- replayEvaluateResultEither match {
+                          case Right(EvaluateResult(replayCost, replayUserErrors)) =>
+                            if (isFailed != replayUserErrors.nonEmpty)
+                              runtime.reportingSpace
+                                .revertToSoftCheckpoint(softCheckpoint)
+                                .as(
+                                  ReplayFailure
+                                    .replayStatusMismatch(isFailed, replayUserErrors.nonEmpty)
+                                    .asLeft[Seq[ReportingEvent]]
+                                )
+                            else if (replayUserErrors.nonEmpty)
+                              runtime.reportingSpace
+                                .revertToSoftCheckpoint(softCheckpoint) >> runtime.reportingSpace.getReport
+                                .map(_.asRight[ReplayFailure])
+                            else if (cost.cost != replayCost.value)
+                              runtime.reportingSpace
+                                .revertToSoftCheckpoint(softCheckpoint)
+                                .as(
+                                  ReplayFailure
+                                    .replayCostMismatch(deploy, cost.cost, replayCost.value)
+                                    .asLeft[Seq[ReportingEvent]]
+                                )
+                            else
+                              runtime.reportingSpace.checkReplayData().attempt >>= {
+                                case Right(_) =>
+                                  runtime.reportingSpace.getReport.map(_.asRight[ReplayFailure])
+                                case Left(throwable) =>
+                                  runtime.reportingSpace.revertToSoftCheckpoint(softCheckpoint) >> {
+                                    throwable match {
+                                      case replayException: ReplayException =>
+                                        ReplayFailure
+                                          .unusedCOMMEvent(deploy, replayException)
+                                          .asLeft[Seq[ReportingEvent]]
+                                          .pure[F]
+                                      case _ =>
+                                        ReplayFailure
+                                          .internalError(deploy, throwable)
+                                          .asLeft[Seq[ReportingEvent]]
+                                          .pure[F]
+                                    }
+                                  }
+                              }
+                          case Left(throwable) =>
+                            runtime.reportingSpace
+                              .revertToSoftCheckpoint(softCheckpoint)
```

Comment:
It would seem that we `revertToSoftCheckpoint` iff we're returning a `ReplayFailure(...).asLeft`. I think it'd make sense to de-duplicate the revert code and make the correlation explicit, so we don't forget about it should there be more error ReplayFailure scenarios.